apc_mutex.c

```diff
@@ -0,0 +1,72 @@
+/*
+  +----------------------------------------------------------------------+
+  | APCu                                                                 |
+  +----------------------------------------------------------------------+
+  | Copyright (c) 2013 The PHP Group                                |
+  +----------------------------------------------------------------------+
+  | This source file is subject to version 3.01 of the PHP license,      |
+  | that is bundled with this package in the file LICENSE, and is        |
+  | available through the world-wide-web at the following url:           |
+  | http://www.php.net/license/3_01.txt                                  |
+  | If you did not receive a copy of the PHP license and are unable to   |
+  | obtain it through the world-wide-web, please send a note to          |
+  | license@php.net so we can mail you a copy immediately.               |
+  +----------------------------------------------------------------------+
+  | Author: Fabian Franz <fabian@lionsad.de>                             |
+  +----------------------------------------------------------------------+
+ */
+#ifdef APC_HAS_PTHREAD_MUTEX
+
+#include "apc_lock.h"
+
+static pthread_mutexattr_t apc_mutex_attr;
+static zend_bool apc_mutex_ready = 0;
+
+PHP_APCU_API zend_bool apc_mutex_init() {
+	if (pthread_mutexattr_init(&apc_mutex_attr) != SUCCESS) {
+		return 0;
+	}
+	if (pthread_mutexattr_setpshared(&apc_mutex_attr, PTHREAD_PROCESS_SHARED) != SUCCESS) {
+		return 0;
+	}
+
+	pthread_mutexattr_settype(&apc_mutex_attr, PTHREAD_MUTEX_RECURSIVE);
+
+	return 1;
+}
+
+/* {{{ Cleanup attributes and statics */
+PHP_APCU_API void apc_mutex_cleanup() {
+	/* once per process please */
+	apc_mutex_ready = 0;
+	pthread_mutexattr_destroy(&apc_mutex_attr);
+}
+
+PHP_APCU_API zend_bool apc_mutex_create(apc_mutex_t *lock) {
+	pthread_mutex_init(lock, &apc_mutex_attr);
+        return 1;
+}
+
+PHP_APCU_API zend_bool apc_mutex_lock(apc_mutex_t *lock) {
+	int result;
+
+	HANDLE_BLOCK_INTERRUPTIONS();
+	if (pthread_mutex_lock(lock) == 0) {
+		return 1;
+	}
+        
+	HANDLE_UNBLOCK_INTERRUPTIONS();
+	apc_warning("Failed to acquire lock");
+	return 0;
+}
+
+PHP_APCU_API zend_bool apc_mutex_unlock(apc_mutex_t *lock) {
+	pthread_mutex_unlock(lock);
+	HANDLE_UNBLOCK_INTERRUPTIONS();
+	return 1;
+}
+
+PHP_APCU_API void apc_mutex_destroy(apc_mutex_t *lock) {
+	// nothing to be done
```

Comment:
Shouldn't this call `pthread_mutex_destroy`?

Reply:
It does not in your lock cleanup, so I guess it's fine?

But yeah, seems like a bug in apc_lock.c. Will fix.

Reply:
fixed, still need to push that